clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -335,6 +335,15 @@ public void poll(long timeout) throws IOException {
         // we use the time at the end of select to ensure that we don't close any connections that
         // have just been processed in pollSelectionKeys
         maybeCloseOldestConnection(endSelect);
+
+        Iterator<String> connectedIt = connected.iterator();
```

Comment:
It might be a little cleaner to avoid the inconsistency in the first place rather than fixing it after the fact. Have you considered calling `connected.remove(channel)` when we add the channel to `disconnected` in `doClose`?